### PR TITLE
Chapter1

### PR DESCRIPTION
**Pull request by moronvv**
### Solutions for Chapter 1

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -489,9 +491,9 @@ Implement a function that returns the last digit of a given number.
   results. Or you can try to guess the function name, search for it and check
   whether it works for you!
 -}
--- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
 
+lastDigit :: Int -> Int
+lastDigit n = (abs n) `mod` 10
```
Comment (vrom911) [inline src/Chapter1.hs]: Nice work on using infix form of calling functions! We don't explain this syntax in our course to make it short. But well done :+1: 

Also, you can remove `()`, they are redundant in this case :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Chapter1.hs [hunk 2]
```diff
@@ -568,8 +575,11 @@ True
 >>> isVowel 'x'
 False
 -}
-isVowel c = error "isVowel: not implemented!"
 
+isVowel :: Char -> Bool
+isVowel c
+  | c `elem` "aeiouy" = True
```
Comment (vrom911) [inline src/Chapter1.hs]: Nice one! 👍🏼 
As `elem` returns `Bool` itself, you can just return the result of this function:

```haskell
isVowel c = c `elem` "aeiouy"
```

One note in here, that sometimes, `elem` could be slower than the explicit pattern matching. I remember there were some benchmarks on one particular case, that showed how moving to pattern matching on each case separately drastically decrease time 🐎

File: src/Chapter1.hs [hunk 3]
```diff
@@ -632,8 +642,12 @@ Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
 
-sumLast2 n = error "sumLast2: Not implemented!"
-
+sumLast2 :: Int -> Int
+sumLast2 n = penultimateNum + lastNum
+  where
+    lastNum = absN `mod` 10
+    penultimateNum = (absN `div` 10) `mod` 10
```
Comment (vrom911) [inline src/Chapter1.hs]: 👍🏼 

In here you use both:

```haskell
absN `mod` 10
absN `div` 10
```

One hint to make your solution even shorter: the standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that is actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(a, b) = absN `divMod` 10
```

You can see how we could pattern match on the pair 🙂

File: src/Chapter1.hs [hunk 4]
```diff
@@ -653,8 +667,13 @@ You need to use recursion in this task. Feel free to return to it later, if you
...
 
-firstDigit n = error "firstDigit: Not implemented!"
-
+firstDigit :: Int -> Int
+firstDigit n
+  | quotient == 0 = absN
+  | otherwise     = firstDigit quotient
+  where
+    absN = abs n
+    quotient = absN `div` 10
```
Comment (vrom911) [inline src/Chapter1.hs:681]: That's a great usage of the `where` construction 👍🏼

**Review by vrom911 — APPROVED** [review]
Nice work on solving our challenges!
And well done with the Advanced Task :medal_sports: 

I left my review and some comments. Hope you find them helpful :slightly_smiling_face: 

Also, I see that you haven't enabled GitHub Actions for your fork. It might be a good idea to do so, and you will be able to see passed and failed test with each PR and new commits.

**Comment by moronvv** [comment]
> Nice work on solving our challenges!
> And well done with the Advanced Task 🏅
> 
> I left my review and some comments. Hope you find them helpful 🙂
> 
> Also, I see that you haven't enabled GitHub Actions for your fork. It might be a good idea to do so, and you will be able to see passed and failed test with each PR and new commits.

Hello. Thanks for the code review! I'll apply all suggestion you have pointed me.

About GitHub Actions. I have all actions allowed in my setting. Should I do anything else to make CI work?
image

**Comment by vrom911** [comment]
@MoronVV I see that it works now, and Chapter One is all green 👍🏼 Well done 👏🏼 
This PR is ready for merge 🚢

**Comment by MoronVV** [comment]
> > Nice work on solving our challenges!
> > And well done with the Advanced Task 🏅
> > I left my review and some comments. Hope you find them helpful 🙂
> > Also, I see that you haven't enabled GitHub Actions for your fork. It might be a good idea to do so, and you will be able to see passed and failed test with each PR and new commits.
> 
> Hello. Thanks for the code review! I'll apply all suggestion you have pointed me.
> 
> About GitHub Actions. I have all actions allowed in my setting. Should I do anything else to make CI work?
> image

Nevermind. Found it! But have some errors for other chapters. Is it ok?

**Pull request merged**